src/GovUk.Education.ExploreEducationStatistics.Common/Services/FileTypeService.cs

```diff
@@ -65,13 +95,100 @@ public bool HasMatchingEncodingType(Stream stream, IEnumerable<string> encodingT
         }
 
+        public async Task<bool> IsValidCsvDataOrMetaFile(
+            Func<Task<Stream>> streamProvider, 
+            string filename)
+        {
+            _logger.LogDebug("Validating that {FileName} has a CSV mime type", filename);
+
+            await using var sampleLinesStream = await GetSampleLinesStream(streamProvider, CsvMimeTypeSampleLineCount);
+
+            if (sampleLinesStream == null)
+            {
+                _logger.LogError("Unable to get sample lines for checking the mime type of {FileName}", filename);
+                return false;
+            }
+            
+            if (!await HasMatchingMimeType(sampleLinesStream, AllowedCsvMimeTypes))
+            {
+                _logger.LogDebug("{FileName} does not have a valid CSV mime type", filename);
+                return false;
+            }
+
+            _logger.LogDebug("{FileName} has a valid CSV mime type", filename);
+
+            _logger.LogDebug("Validating that {FileName} has a valid CSV character encoding", filename);
+
+            await using var encodingStream = await streamProvider.Invoke();
+
+            if (!HasMatchingEncodingType(encodingStream, AllowedCsvEncodingTypes))
+            {
+                _logger.LogDebug("{FileName} does not have a valid CSV content encoding", filename);
+                return false;
+            }
+
+            _logger.LogDebug("{FileName} has a valid CSV content encoding", filename);
+            
+            return true;
+        }
+
+        public Task<bool> IsValidCsvDataOrMetaFile(IFormFile file)
+        {
+            return IsValidCsvDataOrMetaFile(() => Task.FromResult(file.OpenReadStream()), file.FileName);
+        }
+
+        /// <summary>
+        /// Obtain a set of sample lines of the given file, for the purposes of checking the file's mime type and
+        /// content encoding.
+        /// </summary>
+        private async Task<Stream?> GetSampleLinesStream(Func<Task<Stream>> fileStreamProvider, int sampleLineCount)
+        {
+            using var streamReader = new StreamReader(await fileStreamProvider.Invoke());
+
+            var lines = new List<string>();
+            
+            try
+            {
+                var linesRead = 0;
+
+                while (linesRead < sampleLineCount && !streamReader.EndOfStream)
+                {
+                    var nextLine = await streamReader.ReadLineAsync();
+
+                    if (nextLine == null)
+                    {
+                        _logger.LogError("Unable to read next sample line {LineNumber} from CSV", linesRead + 1);
+                        break;
+                    }
+                    
+                    lines.Add(nextLine);
+                    linesRead++;
+                }
+                
+            }
+            catch (Exception e)
+            {
+                _logger.LogError("Unable to read sample lines from CSV - {ErrorMessage}", e.Message);
```

Comment:
I think `_logger.LogError(e, "Unable to read sample lines from CSV")` would be sufficient here, rather than two separate errors which will be recorded in App Insights as two errors.

The message and stacktrace should still be available in the log event that gets recorded.